dask/dataframe/shuffle.py

```diff
@@ -71,7 +71,8 @@ def set_index(df, index, npartitions=None, shuffle=None, compute=False,
                 divisions = [divisions[i] for i in indexes]
 
         if (mins == sorted(mins) and maxes == sorted(maxes) and
```

Comment:
Yes, this section is seeing if the index is *mostly* sorted, that all index values of one partition are less than all index values of the next.  Currently this is all that we guarantee with partitions.  See https://github.com/dask/dask/pull/2150 for a proposed change.